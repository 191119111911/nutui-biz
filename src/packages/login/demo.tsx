import React, { useState, useEffect } from "react";
import { Login } from "./login";
import { useTranslate } from "../../sites/assets/locale";
import { Radio, Cell, CellGroup, Input } from "@nutui/nutui-react";
const { RadioGroup } = Radio;
interface T {
  basic: string;
  hasProtocol: string;
  hasCustom: string;
  hasCustomInput: string;
  hasError: string;
  placeholder: string;
  codeText: string;
  loginText: string;
  errorText1: string;
  errorText2: string;
  titleText: string;
  zhuce: string;
}
const LoginDemo = () => {
  const [translated] = useTranslate<T>({
    "zh-CN": {
      basic: "基本用法",
      hasProtocol: "有知情同意勾选项",
      hasCustom: "用户自定义登录框",
      hasError: "错误提示",
      hasCustomInput: "用户自定义输入框",
      placeholder: "请输入验证码",
      codeText: "获取验证码",
      loginText: "登录",
      errorText1: "账号不存在",
      errorText2: "请输入6位密码",
      titleText: "卡号登录",
      zhuce: "新用户注册",
    },
    "en-US": {
      basic: "Basic usage",
      hasProtocol: "Has informed consent",
      hasCustom: "Custom Login",
      hasCustomInput: "Custom Input",
      hasError: "Error Prompt",
      placeholder: "Please enter the verify code",
      codeText: "Get code",
      loginText: "Login",
      errorText1: "Account does not exist",
      errorText2: "Please enter a 6-digit password",
      titleText: "Login title",
      zhuce: "New User Registration",
    },
  });

  const logoImg =
    "https://img10.360buyimg.com/imagetools/jfs/t1/187998/28/32123/16333/63e346b8F0bff354b/c95da99ea108c463.png";
  const [formParams, setformParams] = useState({
    account: "",
    accountErrorText: "",
    telOrMail: "",
    telOrMailErrorText: "",
    password: "",
    passwordErrorText: "",
    verify: "",
    verifyButtonText: translated.codeText,
  });
  const [formParams2, setformParams2] = useState({
    account: "",
    accountErrorText: "",
    telOrMail: "",
    telOrMailErrorText: "",
    password: "",
    passwordErrorText: "",
    verify: "",
    isShowPwdInput: false,
  });
  const [formParams3, setformParams3] = useState({
    account: "12345",
    accountErrorText: translated.errorText1,
    telOrMail: "",
    telOrMailErrorText: "",
    password: "123",
    passwordErrorText: translated.errorText2,
    verify: "",
  });
  const [getVerify, setGetVerify] = useState(false);
  const [isDisable, setIsDisable] = useState(true);
  const [account, setAccount] = useState("");
  const [customInput, setCustomInput] = useState("");
  const onChange = (value: any, tag: string) => {
    console.log(tag, value);
    if (tag === "account") {
      setAccount(value);
    }
  };
  const queryVerifyCode = (formData: any) => {
    setGetVerify(false);
    console.log("getcode", formData);
    //异步获取校验码成功
    setTimeout(() => {
      setGetVerify(true);
    }, 300);
  };
  const queryLogin = (formData: any) => {
    console.log("login", formData);
  };
  const [radioVal, setRadioVal] = useState("1");
  const handleChange = (v: any) => {
    setRadioVal(v);
    setGetVerify(false);
  };

  useEffect(() => {
    if (radioVal === "5") {
      if (account.length > 0 && customInput.length > 0) {
        setIsDisable(false);
      } else {
        setIsDisable(true);
      }
    }
  }, [account, customInput]);

  return (
    <div className="demo">
      <CellGroup>
        <Cell>
          <RadioGroup
            value={radioVal}
            onChange={handleChange}
            style={{ flexFlow: "wrap" }}
          >
            <Radio value="1">{translated.basic}</Radio>
            <Radio value="2">{translated.hasProtocol}</Radio>
            <Radio value="3">{translated.hasError}</Radio>
            <Radio value="4">{translated.hasCustom}</Radio>
            <Radio value="5">{translated.hasCustomInput}</Radio>
          </RadioGroup>
        </Cell>
      </CellGroup>

      {radioVal === "1" ? (
        <>
          <h2>{translated.basic}</h2>
          <Login
            formParams={formParams}
            loginType="verify"
            logo={logoImg}
            onInputChange={onChange}
            isGetCode={getVerify}
            loginButtonText={translated.loginText}
            onVerifyBtnClick={queryVerifyCode}
            onLoginBtnClick={queryLogin}
            countDownTime={30}
            onForgetBtnClick={() => {
              console.log("点击忘记密码");
            }}
          />
        </>
      ) : null}
      {radioVal === "2" ? (
        <>
          <h2>{translated.hasProtocol}</h2>
          <Login
            formParams={formParams}
            loginType="verify"
            logo={logoImg}
            onInputChange={onChange}
            isGetCode={getVerify}
            onVerifyBtnClick={queryVerifyCode}
            countDownTime={30}
            loginButtonText={translated.loginText}
            slotProtocolText={
              <div>
                勾选后代表您已阅读并同意
                <span style={{ color: "red" }}>《用户隐私政策》</span>
              </div>
            }
          />
        </>
      ) : null}
      {radioVal === "3" ? (
        <>
          <h2>{translated.hasError}</h2>
          <Login
            formParams={formParams3}
            logo={logoImg}
            loginType="pwd"
            onInputChange={onChange}
            loginButtonText={translated.loginText}
          />
        </>
      ) : null}
      {radioVal === "4" ? (
        <>
          <h2>{translated.hasCustom}</h2>
          <Login
            formParams={formParams2}
            title={translated.titleText}
            loginType="pwd"
            onInputChange={onChange}
            isHideSwitchBtn={true}
            loginButtonText={translated.loginText}
            slotBottom={
              <div
                style={{
                  color: "#006FFF",
                  textAlign: "center",
                  fontSize: "14px",
                }}
              >
                {translated.zhuce}
              </div>
            }
          />
        </>
      ) : null}

      {radioVal === "5" ? (
        <>
          <h2>{translated.hasCustomInput}</h2>
          {isDisable}
          <Login
            formParams={formParams2}
            title={translated.titleText}
            loginType="pwd"
            onInputChange={onChange}
            isHideSwitchBtn={true}
            loginButtonDisable={isDisable}
            loginButtonText={translated.loginText}
            onInputClear={(tag) => {
              if (tag === "account") {
                setAccount("");
              }
            }}
            slotInput={
              <div className={`input-wrap`}>
                <div className="input-item">
                  <Input
                    className="nut-input-text"
                    border={false}
                    defaultValue={customInput}
                    placeholder={translated.placeholder}
                    type="text"
                    clearable
                    onChange={(value) => {
                      setCustomInput(value);
                    }}
                    onClear={() => {
                      setCustomInput("");
                    }}
                  />
                  <div className="code-box">
                    <img
                      style={{ width: "65px", height: "30px" }}
                      src="https://img12.360buyimg.com/imagetools/jfs/t1/211415/19/9275/14512/61924b82E09366437/cc5cc7297b9073ae.jpg"
                    />
                  </div>
                </div>
              </div>
            }
          />
        </>
      ) : null}
    </div>
  );
};

export default LoginDemo;
