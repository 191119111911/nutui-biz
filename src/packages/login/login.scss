.nut-login{
    display: flex;
    flex-direction: column;
    padding: 18px ;
    background: $white;
    &__logo{
        width: 80px;
        height: 80px;
        margin: 20px auto 15px;
        img{
            width: 100%;
            height: 100%;
        }
    }
    &__title{
        font-size: 24px;
        font-weight: bold;
        margin: 10px;
    }
    &__content{
        margin: 20px 0px 10px;
        .code-box{
            display: flex;
            align-items: center;
            font-size: 14px;
            min-width:100px;
            padding:0 5px;
            color:#666;
            cursor: pointer;
            &::before{
                content: '';
                display: block;
                width: 1px;
                height: 26px;
                margin-right: 10px;
                background: #e3e3e3;
            }
            &.disabled{
                color: #ccc;
                cursor: not-allowed;
            }
            .counts{
                width: 100%;
                text-align: center;
            }
        }
        .input-wrap{
            display: flex;
            align-items: center;
            position: relative;
            margin-bottom: 0px;
            transition: all 0.2s;
            &.error{
                margin-bottom: 10px;
            }
            .forget-pwd{
                width: 55px;
                font-size: 12px;
                color: #006FFF;
                margin: 0 0 15px 10px;
                cursor: pointer;
            }
            .error-msg{
                position: absolute;
                bottom: -4px;
                font-size: 12px;
                color: red;
                padding: 0 18px;
            }
        }
        .input-item{
            display: flex;
            align-items: center;
            flex:1;
            min-height: 45px;
            border: 1px solid #e3e3e3;
            margin-bottom: 15px;
            border-radius: 4px;
            overflow: hidden;
            .nut-input-text{
                padding:8px 15px;
            }
            .input-text{color: #333;}
            ::-webkit-input-placeholder{
                color: #999;
            }
        }
        .pwd-hide-icon{
            width: 10%;
            display: flex;
            align-items: center;
            margin-right: 10px;
        }
    }
    
    .login-protocal{
        font-size: 12px;
        margin-bottom: 10px;
        .nut-checkbox__label {
            margin-left: 5px;
            font-size: 12px;
        }
    }
    &__btn{
        margin: 0 0 15px;
        border-radius: 4px;
        overflow: hidden;
        .nut-button{
            height: 45px;
            font-size: 16px;
        }
        .nut-button--disabled{
            background: #D8D8D8;
            .nut-button__warp{
                color: #999;
            }
        }
        
    }
    .switch-type{
        color: #006FFF;
        font-size: 14px;
        margin: 0 auto 10px;
        cursor: pointer;
    }
}