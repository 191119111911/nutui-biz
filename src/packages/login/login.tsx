import React, {
  FunctionComponent,
  useEffect,
  useState,
  useRef,
  ReactNode,
} from "react";
import { useConfig } from "@/packages/configprovider";
import classNames from "classnames";
import bem from "@/utils/bem";
import { IComponent } from "@/utils/typings";
import { numericProp } from "@/utils/props";
import {
  Input,
  Button,
  ButtonProps,
  Icon,
  Checkbox,
  Toast,
} from "@nutui/nutui-react";

export type loginType = "verify" | "pwd";
export type showErrorType = "toast" | "bottomMsg";

export interface LoginParamsProps {
  account?: string;
  accountPlaceholder?: string;
  accountErrorText?: string;
  telOrMail?: string | undefined;
  telOrMailPlaceholder?: string;
  telOrMailErrorText?: string;
  password?: string;
  passwordPlaceholder?: string;
  passwordErrorText?: string;
  isShowPwdInput?: boolean;
  verify?: string;
  verifyPlaceholder?: string;
  verifyButtonText?: string;
  verifyErrorText?: string;
  getCodeErrorToast?: string;
  switchLoginText1?: string;
  switchLoginText2?: string;
  forgetPwdText?: string;

  [key: string]: any;
}

export interface LoginFormProps {
  account?: string;
  password?: string;
  telOrMail?: string;
  verify?: string;
  [key: string]: any;
}
export interface LoginProps extends IComponent {
  logo?: string;
  title?: string;
  formParams: LoginParamsProps;
  loginType: loginType;
  loginButtonDisable?: boolean;
  loginButtonText?: string;
  hasForgetPassWord?: boolean;
  showErrorType?: showErrorType;
  toastErrorText?: string;
  hasHidePwd?: boolean;
  isGetCode?: boolean;
  countDownTime?: number | undefined;
  isHideSwitchBtn?: boolean;
  slotProtocolText?: ReactNode;
  slotInput?: ReactNode;
  slotBottom?: ReactNode;
  buttonProps?: ButtonProps;
  onInputChange?: (value: string, tag: string) => void;
  onLoginBtnClick?: (
    formData: LoginFormProps,
    totalData: LoginParamsProps
  ) => void;
  onVerifyBtnClick?: (
    formData: LoginFormProps,
    totalData: LoginParamsProps
  ) => void;
  onForgetBtnClick?: () => void;
  onInputClear?: (tag: string) => void;
  onLoginTypeClick?: () => void;
}

const defaultProps = {
  logo: "",
  title: "",
  formParams: {},
  loginType: "verify",
  loginButtonDisable: true,
  loginButtonText: "登录",
  isGetCode: false,
  hasForgetPassWord: true,
  hasHidePwd: true,
  isHideSwitchBtn: false,
  showErrorType: "toast",
  countDownTime: 60,
} as LoginProps;

export const Login: FunctionComponent<Partial<LoginProps>> = (props) => {
  const { locale } = useConfig();
  const {
    className,
    style,
    logo,
    title,
    formParams,
    loginType,
    loginButtonDisable,
    loginButtonText,
    hasForgetPassWord,
    hasHidePwd,
    isGetCode,
    slotProtocolText,
    slotBottom,
    countDownTime,
    isHideSwitchBtn,
    showErrorType,
    toastErrorText,
    slotInput,
    buttonProps,
    onInputChange,
    onLoginBtnClick,
    onVerifyBtnClick,
    onForgetBtnClick,
    onInputClear,
    onLoginTypeClick,
  } = {
    ...defaultProps,
    ...props,
  };
  let [countTime, setCountTime] = useState(60);
  const [inCountDown, setInCountDown] = useState(false);
  const [isHidePwd, setIsHidePwd] = useState(true);
  const [isLoginDisable, setIsLoginDisable] = useState(loginButtonDisable);
  const [currLoginType, setCurrLoginType] = useState(loginType);
  const [isProtocol, setIsPrtocal] = useState(false);
  let timer: any = useRef(null);
  const [loginForm, setLoginForm] = useState<LoginFormProps>({});

  useEffect(() => {
    //初始化数据
    setLoginParams({ ...loginParams, ...formParams });

    setLoginForm({
      account: formParams.account,
      password: formParams.password,
      telOrMail: formParams.telOrMail,
      verify: formParams.verify,
    });
  }, [formParams]);

  useEffect(() => {
    countDownTime && !inCountDown && setCountTime(countDownTime);
  }, [countDownTime]);

  //监听登录按钮是否禁用
  useEffect(() => {
    setIsLoginDisable(loginButtonDisable);
  }, [loginButtonDisable]);

  const [loginParams, setLoginParams] = useState<LoginParamsProps>({
    account: "",
    accountPlaceholder: locale.login.accountPlaceholder,
    accountErrorText: "",
    telOrMail: "",
    telOrMailPlaceholder: locale.login.telOrMailPlaceholder,
    telOrMailErrorText: "",
    password: "",
    passwordPlaceholder: locale.login.passwordPlaceholder,
    passwordErrorText: "",
    isShowPwdInput: true,
    verify: "",
    verifyPlaceholder: locale.login.verifyPlaceholder,
    verifyButtonText: locale.login.verifyButtonText,
    verifyErrorText: "",
    getCodeErrorToast: locale.login.getCodeErrorToast,
    switchLoginText1: locale.login.switchLoginText1,
    switchLoginText2: locale.login.switchLoginText2,
    forgetPwdText: locale.login.forgetPwdText,
  });

  const switchLogin = () => {
    setCurrLoginType(currLoginType === "pwd" ? "verify" : "pwd");
    resetParams();
    onLoginTypeClick && onLoginTypeClick();
  };
  //重置
  const resetParams = () => {
    const clearObj = {
      account: "",
      accountErrorText: "",
      telOrMail: "",
      telOrMailErrorText: "",
      password: "",
      passwordErrorText: "",
      verify: "",
      verifyErrorText: "",
    };
    setLoginForm({
      account: "",
      password: "",
      telOrMail: "",
      verify: "",
    });
    setLoginParams({ ...loginParams, ...clearObj });
    setIsPrtocal(false);
    clearInterval(timer.current);
    setInCountDown(false);
    countDownTime && setCountTime(countDownTime);
  };
  useEffect(() => {
    const { account, telOrMail, password, verify, isShowPwdInput } =
      loginParams;

    //登录状态可点击情况
    let isPwdNotEmpty =
      currLoginType === "pwd" &&
      account != "" &&
      password != "" &&
      isShowPwdInput;
    let isAccountNotEmpty =
      currLoginType === "pwd" && account != "" && !isShowPwdInput;
    let isVerfiyNotEmpty =
      currLoginType === "verify" && telOrMail != "" && verify != "";
    // 用户自定义输入框slotInput时，登录按钮是否可点击用户控制
    if (!slotInput) {
      if (slotProtocolText) {
        if (
          (isPwdNotEmpty || isAccountNotEmpty || isVerfiyNotEmpty) &&
          isProtocol
        ) {
          setIsLoginDisable(false);
          return;
        }
      } else {
        if (isPwdNotEmpty || isAccountNotEmpty || isVerfiyNotEmpty) {
          setIsLoginDisable(false);
          return;
        }
      }
    }

    !slotInput && setIsLoginDisable(true);
  }, [loginParams, slotProtocolText, isProtocol, slotInput]);
  //toast 错误提示监听
  useEffect(() => {
    if (showErrorType === "toast" && toastErrorText?.length) {
      Toast.text(toastErrorText);
    }
  }, [toastErrorText]);

  const inputOnChange = (value: any, tag: string) => {
    let params: LoginParamsProps = { ...loginParams };
    let form: LoginFormProps = { ...loginForm };
    params[tag] = value;
    form[tag] = value;
    setLoginParams({ ...loginParams, ...params });
    setLoginForm({ ...loginForm, ...form });
    onInputChange && onInputChange(value, tag);
  };

  // 获取验证码
  const getCode = () => {
    const { telOrMail, getCodeErrorToast } = loginParams;
    //校验手机号和邮箱
    if (telOrMail?.length) {
      onVerifyBtnClick && onVerifyBtnClick(loginForm, loginParams);
    } else {
      Toast.text(getCodeErrorToast, { duration: 2 });
    }
  };
  //异步获取验证码，控制倒计时
  useEffect(() => {
    if (isGetCode && countDownTime && !inCountDown) {
      setCountTime(countDownTime);
      countDown(countDownTime);
    }
  }, [isGetCode]);

  // 倒计时
  const countDown = (time: number) => {
    setCountTime(countTime--);
    setInCountDown(true);
    timer.current = setInterval(() => {
      setCountTime(countTime--);
      if (countTime < 10) {
        setCountTime(countTime);
        if (countTime == 0) {
          clearInterval(timer.current);
          setInCountDown(false);
          setCountTime(time);
        }
      }
    }, 1000);
  };

  const forgetClick = () => {
    onForgetBtnClick && onForgetBtnClick();
  };

  const loginClick = () => {
    onLoginBtnClick && onLoginBtnClick(loginForm, loginParams);
  };
  const inputClear = (tag: string) => {
    let params: LoginParamsProps = loginParams;
    let form: LoginFormProps = { ...loginForm };
    params[tag] = "";
    form[tag] = "";
    onInputClear && onInputClear(tag);
    setLoginParams({ ...loginParams, ...params });
    setLoginForm({ ...loginForm, ...form });
  };

  const isError = (tag: string) => {
    let name = tag + "ErrorText";
    return loginParams[name] != "" && showErrorType === "bottomMsg";
  };
  //隐藏密码icon
  const pwdEyesIcon = () => {
    return (
      <div
        className={`${b("hide-icon")}`}
        onClick={() => {
          setIsHidePwd(!isHidePwd);
        }}
      >
        <Icon
          name={isHidePwd ? "marshalling" : "eye"}
          size="14"
          color={isHidePwd ? "#ccc" : "#666"}
        />
      </div>
    );
  };
  //获取验证码倒计时按钮
  const countDownTpl = () => {
    return !inCountDown ? (
      <div className={`${b("code-box")}`} onClick={getCode}>
        {loginParams.verifyButtonText}
      </div>
    ) : (
      <div className={classNames([b("code-box"), "disabled"])}>
        <div className="counts">{countTime}s</div>
      </div>
    );
  };
  const inputTpl = (tag: string) => {
    let placeholder = tag + "Placeholder";
    let errorText = tag + "ErrorText";

    return (
      <div
        className={classNames([b("input-wrap"), isError(tag) ? "error" : ""])}
      >
        <div className={`${b("input-item")}`}>
          <Input
            className="nut-input-text"
            border={false}
            defaultValue={loginParams[tag]}
            name={tag}
            placeholder={loginParams[placeholder]}
            type={isHidePwd && tag === "password" ? "password" : "text"}
            clearable
            onChange={(e) => {
              inputOnChange(e, tag);
            }}
            onClear={() => {
              inputClear(tag);
            }}
          />
          {tag === "password" && hasHidePwd && pwdEyesIcon()}
          {tag === "verify" && countDownTpl()}
        </div>
        {tag === "password" && hasForgetPassWord && (
          <div className="forget-pwd" onClick={forgetClick}>
            {loginParams.forgetPwdText}
          </div>
        )}
        {loginParams[errorText] && (
          <div className="error-msg">{loginParams[errorText]}</div>
        )}
      </div>
    );
  };

  const b = bem("login");
  return (
    <div className={classNames([b(), className])} style={style}>
      {logo && (
        <div className={`${b("logo")}`}>
          <img src={logo} />
        </div>
      )}
      {title && <div className={`${b("title")}`}>{title}</div>}
      <div className={`${b("content")}`}>
        {currLoginType == "pwd" ? (
          <>
            {inputTpl("account")}
            {loginParams.isShowPwdInput && inputTpl("password")}
          </>
        ) : (
          <>
            {inputTpl("telOrMail")}
            {inputTpl("verify")}
          </>
        )}
        {slotInput}
        {slotProtocolText && (
          <div className={`${b("protocal")}`}>
            <Checkbox
              iconSize={14}
              checked={isProtocol}
              onChange={(state) => {
                setIsPrtocal(state);
              }}
            ></Checkbox>
            <div className="customer-protocal">{slotProtocolText}</div>
          </div>
        )}
      </div>
      <div className={`${b("btn")}`}>
        <Button
          block
          type="danger"
          shape="square"
          disabled={isLoginDisable}
          onClick={loginClick}
          {...buttonProps}
        >
          {loginButtonText ? loginButtonText : locale.login.loginButtonText}
        </Button>
      </div>
      {!isHideSwitchBtn && (
        <div className={`${b("switch-type")}`} onClick={switchLogin}>
          {currLoginType === "verify"
            ? loginParams.switchLoginText1
            : loginParams.switchLoginText2}
        </div>
      )}
      {slotBottom && <div className="custom-slot">{slotBottom}</div>}
    </div>
  );
};

Login.defaultProps = defaultProps;
Login.displayName = "NutLogin";
