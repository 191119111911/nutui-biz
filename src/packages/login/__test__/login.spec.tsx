import React, { useState } from "react";
import { render, waitFor, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Login } from "../login";

const logoImg =
  "https://img10.360buyimg.com/imagetools/jfs/t1/187998/28/32123/16333/63e346b8F0bff354b/c95da99ea108c463.png";
const formParams = {
  account: "",
  accountPlaceholder: "请输入账号",
  accountErrorText: "",
  telOrMail: "",
  telOrMailPlaceholder: "请输入手机号或邮箱",
  telOrMailErrorText: "",
  password: "",
  passwordPlaceholder: "请输入账号密码",
  passwordErrorText: "",
  verifyCode: "",
  verifyPlaceholder: "请输入验证码",
  verifyButtonText: "获取验证码",
};
const getVerify = false;

test("Verify Login test", async () => {
  const { container } = render(
    <Login
      formParams={formParams}
      loginType="verify"
      logo={logoImg}
      isGetCode={getVerify}
      loginButtonText={"登录"}
      countDownTime={30}
    />
  );
  const logoImge = container.querySelector(".nut-login__logo img");
  const telEl = container.querySelector('.input-text[name="telOrMail"]');
  const verifyEl = container.querySelector('.input-text[name="verifyCode"]');
  const codeBtn = container.querySelector(".code-box") as HTMLElement;
  const loginBtn = container.querySelector(".nut-login__btn .nut-button");
  await waitFor(() => {
    expect(logoImge).toHaveAttribute("src", `${logoImg}`);
    expect(container.querySelector(".switch-type")?.innerHTML).toBe(
      "账号密码登录"
    );
    expect(telEl).toHaveAttribute(
      "placeholder",
      `${formParams.telOrMailPlaceholder}`
    );
    expect(verifyEl).toHaveAttribute(
      "placeholder",
      `${formParams.verifyPlaceholder}`
    );
    expect(codeBtn.innerHTML).toBe(formParams.verifyButtonText);
    expect(loginBtn).toHaveClass("nut-button--disabled");
    expect(container).toMatchSnapshot();
  });
});

test("Password Login test", async () => {
  const { container } = render(
    <Login
      formParams={formParams}
      loginType="pwd"
      logo={logoImg}
      isGetCode={getVerify}
      loginButtonText={"登录"}
      countDownTime={30}
    />
  );
  const accountEl = container.querySelector('.input-text[name="account"]');
  const pwdEl = container.querySelector('.input-text[name="password"]');
  await waitFor(() => {
    // expect(logoImge).toHaveAttribute("src", `${logoImg}`);
    // expect(container.querySelector(".switch-type")?.innerHTML).toBe(
    //   "账号密码登录"
    // );
    expect(accountEl).toHaveAttribute(
      "placeholder",
      `${formParams.accountPlaceholder}`
    );
    expect(pwdEl).toHaveAttribute(
      "placeholder",
      `${formParams.passwordPlaceholder}`
    );
    expect(container.querySelector(".pwd-hide-icon")).toBeTruthy;
    expect(container.querySelector(".forget-pwd")).toBeTruthy;
    expect(container.querySelector(".switch-type")?.innerHTML).toBe(
      "手机/邮箱登录"
    );
    expect(container).toMatchSnapshot();
  });
});

test("Has informed consent", async () => {
  const { container } = render(
    <Login
      formParams={formParams}
      loginType="verify"
      logo={logoImg}
      countDownTime={30}
      slotProtocolText={
        <div>
          勾选后代表您已阅读并同意
          <span style={{ color: "red" }}>《用户隐私政策》</span>
        </div>
      }
    />
  );
  const checkEl = container.querySelector(
    ".login-protocal .nut-checkbox__label "
  );

  await waitFor(() => {
    expect(checkEl?.innerHTML).toBe(
      `<div>勾选后代表您已阅读并同意<span style="color: red;">《用户隐私政策》</span></div>`
    );
  });
});

const formParams2 = {
  telOrMail: "13123456789",
  verifyCode: "1111",
  account: "张三",
  accountErrorText: "账号不存在",
  password: "123456",
  passwordErrorText: "请输入6位密码",
};

test("event test", async () => {
  const onChange = jest.fn();
  const verifyClick = jest.fn();
  const loginClick = jest.fn();
  const { container } = render(
    <Login
      formParams={formParams2}
      loginType="verify"
      logo={logoImg}
      onInputChange={onChange}
      isGetCode={getVerify}
      loginButtonText={"登录"}
      onVerifyBtnClick={verifyClick}
      onLoginBtnClick={loginClick}
      countDownTime={30}
    />
  );
  const telEl = container.querySelector('.input-text[name="telOrMail"]');
  const verifyEl = container.querySelector('.input-text[name="verifyCode"]');
  const loginBtn = container.querySelector(".nut-login__btn .nut-button");
  const codeBtn = container.querySelector(".code-box") as HTMLElement;
  expect(telEl).toHaveAttribute("value", `${formParams2.telOrMail}`);
  expect(verifyEl).toHaveAttribute("value", `${formParams2.verifyCode}`);
  expect(loginBtn).not.toHaveClass("nut-button--disabled");
  codeBtn && fireEvent.click(codeBtn);
  loginBtn && fireEvent.click(loginBtn);
  telEl && fireEvent.change(telEl, { target: { value: "12345" } });

  await waitFor(() => {
    expect(verifyClick).toBeCalled();
    expect(loginClick).toBeCalled();
    waitFor(() => {
      expect(onChange).toBeCalled();
    });
  });
});

test("error test", async () => {
  const onChange = jest.fn();
  const verifyClick = jest.fn();
  const loginClick = jest.fn();
  const { container } = render(
    <Login
      formParams={formParams2}
      loginType="pwd"
      logo={logoImg}
      onInputChange={onChange}
      isGetCode={getVerify}
      loginButtonText={"登录"}
      onVerifyBtnClick={verifyClick}
      onLoginBtnClick={loginClick}
      countDownTime={30}
    />
  );
  const errorMsg = container.querySelectorAll(".error-msg");

  expect(errorMsg[0].innerHTML).toBe(`${formParams2.accountErrorText}`);
  expect(errorMsg[1].innerHTML).toBe(`${formParams2.passwordErrorText}`);
});
