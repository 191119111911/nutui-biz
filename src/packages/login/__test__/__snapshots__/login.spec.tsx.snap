// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Login test 1`] = `
<div>
  <div
    class="nut-login"
  >
    <div
      class="nut-login__logo"
    >
      <img
        src="https://img10.360buyimg.com/imagetools/jfs/t1/187998/28/32123/16333/63e346b8F0bff354b/c95da99ea108c463.png"
      />
    </div>
    <div
      class="nut-login__content"
    >
      <div
        class="input-wrap "
      >
        <div
          class="input-item"
        >
          <div
            class="nut-input  nut-input-text"
          >
            <div
              class="nut-input-value"
            >
              <div
                class="nut-input-main-con"
              >
                <div
                  class="nut-input-inner"
                >
                  <input
                    class="input-text"
                    maxlength="9999"
                    name="account"
                    placeholder="请输入账号"
                    style="text-align: left;"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="input-wrap "
      >
        <div
          class="input-item"
        >
          <div
            class="nut-input  nut-input-text"
          >
            <div
              class="nut-input-value"
            >
              <div
                class="nut-input-main-con"
              >
                <div
                  class="nut-input-inner"
                >
                  <input
                    class="input-text"
                    maxlength="9999"
                    name="password"
                    placeholder="请输入账号密码"
                    style="text-align: left;"
                    type="password"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="pwd-hide-icon"
          >
            <i
              class="nutui-iconfont nut-icon nut-icon-marshalling "
              style="color: rgb(204, 204, 204); font-size: 14px; height: 14px; width: 14px;"
            />
          </div>
        </div>
        <div
          class="forget-pwd"
        >
          忘记密码
        </div>
      </div>
    </div>
    <div
      class="nut-login__btn"
    >
      <button
        class="nut-button nut-button--danger nut-button--normal nut-button--square nut-button--block nut-button--disabled "
      >
        <div
          class="nut-button__warp"
        >
          <div
            class=""
          >
            登录
          </div>
        </div>
      </button>
    </div>
    <div
      class="switch-type"
    >
      手机/邮箱登录
    </div>
  </div>
</div>
`;

exports[`Verify Login test 1`] = `
<div>
  <div
    class="nut-login"
  >
    <div
      class="nut-login__logo"
    >
      <img
        src="https://img10.360buyimg.com/imagetools/jfs/t1/187998/28/32123/16333/63e346b8F0bff354b/c95da99ea108c463.png"
      />
    </div>
    <div
      class="nut-login__content"
    >
      <div
        class="input-wrap "
      >
        <div
          class="input-item"
        >
          <div
            class="nut-input  nut-input-text"
          >
            <div
              class="nut-input-value"
            >
              <div
                class="nut-input-main-con"
              >
                <div
                  class="nut-input-inner"
                >
                  <input
                    class="input-text"
                    maxlength="9999"
                    name="telOrMail"
                    placeholder="请输入手机号或邮箱"
                    style="text-align: left;"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="input-wrap "
      >
        <div
          class="input-item"
        >
          <div
            class="nut-input  nut-input-text"
          >
            <div
              class="nut-input-value"
            >
              <div
                class="nut-input-main-con"
              >
                <div
                  class="nut-input-inner"
                >
                  <input
                    class="input-text"
                    maxlength="9999"
                    name="verifyCode"
                    placeholder="请输入验证码"
                    style="text-align: left;"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="code-box"
          >
            获取验证码
          </div>
        </div>
      </div>
    </div>
    <div
      class="nut-login__btn"
    >
      <button
        class="nut-button nut-button--danger nut-button--normal nut-button--square nut-button--block nut-button--disabled "
      >
        <div
          class="nut-button__warp"
        >
          <div
            class=""
          >
            登录
          </div>
        </div>
      </button>
    </div>
    <div
      class="switch-type"
    >
      账号密码登录
    </div>
  </div>
</div>
`;
